feat: add `math/base/special/kernel-log1pf`

Neerajpathak07 opened the pull request:
Progresses #649, Resolves #5412.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `math/base/special/kernel-log1pf` which is the single-precision version of `math/base/special/kernel-log1p`

## Related Issues

> Does this pull request have any related issues?

This pull request:

- progresses #649

-   resolves #5412.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: add kernel-log1pf

Neerajpathak07 committed:
chore: update markdown

Neerajpathak07 commented:
Will be needing this to implement some big packages like `log2f` and `log10f`
Also there are little gaps in the code coverage for `polyval_p.js` and `polyval_q.js` was it because these functions were auto-generated or something that we should expect??
cc: @kgryte @Planeshifter @gunjjoshi

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/lib/polyval_p.js:
```diff
@@ -0,0 +1,47 @@
+* @returns {number} evaluated polynomial
+*/
+function evalpoly( x ) {
+	if ( x === 0.0 ) {
+		return 0.40000972152;
+	}
```
According to the code coverage report, this part seems unreachable here and in `polyval_q.js`. You can backtrack from here to track down if there is/are some value(s) that can be supplied to this, so that this becomes reachable.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.js:
```diff
@@ -0,0 +1,73 @@
+		}
+	}
+	t.end();
+});
```
```suggestion
});

tape( 'the function returns `0` if provided `0`', function test( t ) {
	var v = kernelLog1pf( 0.0 );
	t.equal( v, 0.0, 'returns expected value' );
	t.end();
} );
```

Specifically, you can add a test for `0` similar to this.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.native.js:
```diff
@@ -0,0 +1,82 @@
+		}
+	}
+	t.end();
+});
```
```suggestion
});

tape( 'the function returns `0` if provided `0`', opts, function test( t ) {
	var v = kernelLog1pf( 0.0 );
	t.equal( v, 0.0, 'returns expected value' );
	t.end();
} );
```

Neerajpathak07 replied:
alright I'll try this once and get back to you.

gunjjoshi commented:
@Neerajpathak07 You'll achieve 100% code coverage once you've added the above two tests.

kgryte commented:
In general, it is possible that not allow Horner branches will be exercised. However, in this case, as @gunjjoshi pointed out, you can achieve full coverage by reasoning about the implementation code.

kgryte commented:
@gunjjoshi Would you mind doing a review of this PR?

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+*/
+static double benchmark( void ) {
+	double elapsed;
+	double x[ 100 ];
```
```suggestion
	float x[ 100 ];
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+	double elapsed;
+	double x[ 100 ];
+	double z;
+	double t;
```
```suggestion
	double t;
	float z;
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		z = stdlib_base_kernel_log1pf( x[ i%100 ] );
```
```suggestion
		z = stdlib_base_kernel_log1pf( x[ i % 100 ] );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+
+var bench = require( '@stdlib/bench' );
+var uniform = require( '@stdlib/random/array/uniform' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = kernelLog1pf( x[ i % x.length ] );
+		if ( isnan( y ) ) {
```
```suggestion
		if ( isnanf( y ) ) {
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+		}
+	}
+	b.toc();
+	if ( isnan( y ) ) {
```
```suggestion
	if ( isnanf( y ) ) {
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+	var y;
+	var i;
+
+	x = uniform( 100, 0.70710677, 1.0 );
```
@Neerajpathak07 This range should be `[1/√2, 2/√2]`, as given here: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/kernel-log1p/benchmark/benchmark.js#L39.
Moreover, instead of using this magic constant directly, we can replace it with a package such as `@stdlib/constants/float32/sqrt-half` and `@stdlib/constants/float32/sqrt-halfsqrt-two`.

Neerajpathak07 replied:
the latter for `@stdlib/constants/float32/sqrt-halfsqrt-two` isn't there Ig

Neerajpathak07 replied:
we could use `2 / FLOAT32_SQRT2`

gunjjoshi replied:
We have it: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/constants/float32/sqrt-two

Neerajpathak07 replied:
yeah but it the the single-precision `sqrt` of 2 right so using this `2 / FLOAT32_SQRT2` would be valid.

gunjjoshi replied:
@Neerajpathak07 We can have something like this:

```js
var FLOAT32_SQRT_TWO = require( '@stdlib/constants/float32/sqrt-two' );
var FLOAT32_SQRT_HALF = require( '@stdlib/constants/float32/sqrt-half' );
.
// Rest of the code...
.
x = uniform( 100, FLOAT32_SQRT_HALF, FLOAT32_SQRT_TWO );
```

Neerajpathak07 replied:
updated the benchmark!

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,61 @@
+var resolve = require( 'path' ).resolve;
+var bench = require( '@stdlib/bench' );
+var uniform = require( '@stdlib/random/array/uniform' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,61 @@
+	var y;
+	var i;
+
+	x = uniform( 100, 0.70710677, 1.0 );
```
Same comment regarding the range and the use of constant packages.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,61 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = kernelLog1pf( x[ i % x.length ] );
+		if ( isnan( y ) ) {
```
```suggestion
		if ( isnanf( y ) ) {
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,61 @@
+		}
+	}
+	b.toc();
+	if ( isnan( y ) ) {
```
```suggestion
	if ( isnanf( y ) ) {
```

Neerajpathak07 committed:
chore: achieve 100% code coverage

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

Neerajpathak07 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-kernel-log1pf

Neerajpathak07 commented:
@gunjjoshi Thanks for the assist I  believe we have achieved 100% code coverage.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
```diff
@@ -0,0 +1,203 @@
+
+## Notes
+
+-   This function is a helper function for computing logarithms in base `e`. Argument reduction and adding the final term of the polynomial must be done by the caller for increased accuracy when different bases are used. For reference, see [`@stdlib/math/base/special/log2f`][@stdlib/math/base/special/log2f] and [`@stdlib/math/base/special/log10f`][@stdlib/math/base/special/log10f].
```
@Neerajpathak07 We don't have the implementations for `log2f` and `log10f` yet.

gunjjoshi replied:
Alright, but those implementations would depend on `kernel-log1pf`. I think we can keep this as it is, and implement those two packages once we're done with this.

Neerajpathak07 replied:
I had made draft PR's on those packages which is why I was looking forward to work on this one.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
````diff
@@ -0,0 +1,203 @@
+### Usage
+
+```c
+#include "stdlib/math/base/special/kernel_log1p.h"
````
```suggestion
#include "stdlib/math/base/special/kernel_log1pf.h"
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
````diff
@@ -0,0 +1,203 @@
+Computes `logf(1+f) - f` for `1+f` in `~[sqrtf(2)/2, sqrtf(2)]`.
+
+```c
+double out = stdlib_base_kernel_log1pf( 1.0 );
````
```suggestion
float out = stdlib_base_kernel_log1pf( 1.0f );
```

Neerajpathak07 replied:
had made these changes in the previous commit

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
````diff
@@ -0,0 +1,203 @@
+
+```c
+double out = stdlib_base_kernel_log1pf( 1.0 );
+// returns ~0.1931
````
```suggestion
// returns ~0.1931f
```

Neerajpathak07 replied:
had made these changes in the previous commit

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
```diff
@@ -0,0 +1,203 @@
+    float out;
+    int i;
+    for ( i = 0; i < 10; i++ ) {
+        out = stdlib_base_kernel_log1p( x[ i ] );
```
```suggestion
        out = stdlib_base_kernel_log1pf( x[ i ] );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
```diff
@@ -0,0 +1,203 @@
+
+-   <span class="package-name">[`@stdlib/math/base/special/kernel-log1p`][@stdlib/math/base/special/kernel-log1p]</span><span class="delimiter">: </span><span class="description">Computes `log(1+f) - f` for `1+f` in `~[sqrt(2)/2, sqrt(2)]`.</span>
+-   <span class="package-name">[`@stdlib/math/base/special/log1p`][@stdlib/math/base/special/log1p]</span><span class="delimiter">: </span><span class="description">evaluate the natural logarithm of 1+x.</span>
+
```
```suggestion
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/README.md:
```diff
@@ -0,0 +1,203 @@
+[@stdlib/math/base/special/kernel-log1p]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/kernel-log1p
+
+[@stdlib/math/base/special/log1p]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/log1p
+
```
```suggestion
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/lib/main.js:
```diff
@@ -0,0 +1,83 @@
+
+/**
+* Computes `logf(1+f) - f` for `1+f` in `~[sqrtf(2)/2, sqrtf(2)]`.
+*
```
```suggestion
*
* ## Method
*
* This function is a helper function for computing logarithms in base \\(e\\), and what follows describes the overall strategy for doing so. The argument reduction and adding the final term of the polynomial are done by the caller for increased accuracy when different bases are used.
*
* 1.  Argument Reduction. Find \\(k\\) and \\(f\\) such that
*
*     ```tex
*     x = 2^k \cdot (1+f)
*     ```
*
*     where \\(\sqrt(2)/2 < 1+f < \sqrt(2)\\).
*
* 2.  Approximation of \\(\operatorname{log}(1+f)\\). Let
*
*     ```tex
*     \begin{align*}
*     s &= \frac{f}{2+f} \\
*     &= 2s + \frac{2}{3} s^3 + \frac{2}{5} s^5 + \ldots \\
*     &= 2s + s R
*     \end{align*}
*     ```
*
*     based on
*
*     ```tex
*     \operatorname{log}(1+f) &= \operatorname{log}(1+s) - \operatorname{log}(1-s)
*     ```
*
*     We use a special Reme algorithm on \\(\[0,0.1716]\\) to generate a polynomial of degree \\(14\\) to approximate \\(R\\). The maximum error of this polynomial approximation is bounded by \\(2^{-58.45}\\). In other words,
*
*     ```tex
*     R(z) ~ L_{g1} s^2 + L_{g2} s^4 + L_{g3} s^6 + L_{g4} s^8 + L_{g5} s^{10} + L_{g6} s^{12} + L_{g7} s^{14}
*     ```
*
*     where the values of \\(L_{g1}\\) to \\(L_{g7}\\) are the polynomial coefficients used in the program below and
*
*     ```tex
*     L_{g1} s^2 + \ldots + L_{g7} s^{14} - R(z) \leq 2^{-58.45}
*     ```
*
*     Note that
*
*     ```tex
*     2s = f - s \cdot f = f - h_{fsq} + (s \cdot h_{fsq})
*     ```
*
*     where \\(h_{fsq} = f^{2}/2\\).
*
*     In order to guarantee an error in \\(\operatorname{log}\\) below 1 ulp, we compute \\(\operatorname{log}\\) by
*
*     ```tex
*     \begin{align*}
*     \operatorname{log}(1+f) &= f - s (f - R) & \textrm{(if f is not too large)} \\
*     \operatorname{log}(1+f) &= f - (h_{fsq} - s (h_{fsq}+R)) & \textrm{(better accuracy)}
*     \end{align*}
*
* 3.  Finally,
*
*     ```tex
*     \begin{align*}
*     \operatorname{log}(x) &= k \cdot \operatorname{ln2} + \operatorname{log}(1+f) \\
*     &= k \cdot \operatorname{ln2}_{hi} + (f-(h_{fsq}-(s \cdot (h_{fsq}+R) + k \cdot \operatorname{ln2}_{lo})))
*     \end{align*}
*     ```
*
*     Here, \\(\operatorname{ln2}\\) is split into two floating point numbers:
*
*     ```tex
*     \operatorname{ln2} = \operatorname{ln2}_{hi} + \operatorname{ln2}_{lo}
*     ```
*
*     where \\(n \cdot \operatorname{ln2}_{hi}\\) is always exact for \\(|n| < 2000\\).
*
```

gunjjoshi replied:
Let's preserve the comments from the original source.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/lib/main.js:
```diff
@@ -0,0 +1,83 @@
+	R = t2 + t1;
+	hfsq = 0.5 * f * f;
+	return s * ( hfsq + R );
+}
```
@Neerajpathak07 Let's wrap all the arithmetic operations with `float64ToFloat32` here.
For reference, see: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/acosf/lib/main.js

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/scripts/evalpoly.js:
```diff
@@ -0,0 +1,130 @@
+// VARIABLES //
+
+// Polynomial coefficients ordered in ascending degree...
+var P = [
+	0.40000972152,
+	0.24279078841
+];
+var Q = [
+	0.66666662693,
+	0.28498786688
+];
```
@Neerajpathak07 These polynomial coefficients differ from those in the [double-precision implementation](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/kernel-log1p/scripts/evalpoly.js#L41). Any particular reason for this?

Neerajpathak07 replied:
Yes according to the [msun](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/k_logf.h?revision=367086&view=markup) implementation which was also the reference implementation for `kernel-log1p`

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/src/main.c:
```diff
@@ -0,0 +1,107 @@
+
+/**
+* Computes `logf(1+f) - f` for `1+f` in `~[sqrtf(2)/2, sqrtf(2)]`.
+*
```
```suggestion
*
* ## Method
*
* This function is a helper function for computing logarithms in base \\(e\\), and what follows describes the overall strategy for doing so. The argument reduction and adding the final term of the polynomial are done by the caller for increased accuracy when different bases are used.
*
* 1.  Argument Reduction. Find \\(k\\) and \\(f\\) such that
*
*     ```tex
*     x = 2^k \cdot (1+f)
*     ```
*
*     where \\(\sqrt(2)/2 < 1+f < \sqrt(2)\\).
*
* 2.  Approximation of \\(\operatorname{log}(1+f)\\). Let
*
*     ```tex
*     \begin{align*}
*     s &= \frac{f}{2+f} \\
*     &= 2s + \frac{2}{3} s^3 + \frac{2}{5} s^5 + \ldots \\
*     &= 2s + s R
*     \end{align*}
*     ```
*
*     based on
*
*     ```tex
*     \operatorname{log}(1+f) &= \operatorname{log}(1+s) - \operatorname{log}(1-s)
*     ```
*
*     We use a special Reme algorithm on \\(\[0,0.1716]\\) to generate a polynomial of degree \\(14\\) to approximate \\(R\\). The maximum error of this polynomial approximation is bounded by \\(2^{-58.45}\\). In other words,
*
*     ```tex
*     R(z) ~ L_{g1} s^2 + L_{g2} s^4 + L_{g3} s^6 + L_{g4} s^8 + L_{g5} s^{10} + L_{g6} s^{12} + L_{g7} s^{14}
*     ```
*
*     where the values of \\(L_{g1}\\) to \\(L_{g7}\\) are the polynomial coefficients used in the program below and
*
*     ```tex
*     L_{g1} s^2 + \ldots + L_{g7} s^{14} - R(z) \leq 2^{-58.45}
*     ```
*
*     Note that
*
*     ```tex
*     2s = f - s \cdot f = f - h_{fsq} + (s \cdot h_{fsq})
*     ```
*
*     where \\(h_{fsq} = f^{2}/2\\).
*
*     In order to guarantee an error in \\(\operatorname{log}\\) below 1 ulp, we compute \\(\operatorname{log}\\) by
*
*     ```tex
*     \begin{align*}
*     \operatorname{log}(1+f) &= f - s (f - R) & \textrm{(if f is not too large)} \\
*     \operatorname{log}(1+f) &= f - (h_{fsq} - s (h_{fsq}+R)) & \textrm{(better accuracy)}
*     \end{align*}
*
* 3.  Finally,
*
*     ```tex
*     \begin{align*}
*     \operatorname{log}(x) &= k \cdot \operatorname{ln2} + \operatorname{log}(1+f) \\
*     &= k \cdot \operatorname{ln2}_{hi} + (f-(h_{fsq}-(s \cdot (h_{fsq}+R) + k \cdot \operatorname{ln2}_{lo})))
*     \end{align*}
*     ```
*
*     Here, \\(\operatorname{ln2}\\) is split into two floating point numbers:
*
*     ```tex
*     \operatorname{ln2} = \operatorname{ln2}_{hi} + \operatorname{ln2}_{lo}
*     ```
*
*     where \\(n \cdot \operatorname{ln2}_{hi}\\) is always exact for \\(|n| < 2000\\).
*
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/src/main.c:
```diff
@@ -0,0 +1,107 @@
+* @return     output value
+*
+* @example
+* double out = stdlib_base_kernel_log1p( 1.0 );
```
```suggestion
* float out = stdlib_base_kernel_log1pf( 1.0f );
```

Neerajpathak07 replied:
had made these changes in the previous commit

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/src/main.c:
```diff
@@ -0,0 +1,107 @@
+*
+* @example
+* double out = stdlib_base_kernel_log1p( 1.0 );
+* // returns ~0.1931
```
```suggestion
* // returns ~0.1931f
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.js:
```diff
@@ -0,0 +1,73 @@
+			t.equal( y, e, 'x: '+x[ i ]+'. E: '+e );
+		} else {
+			delta = absf( y - e );
+			tol = 130.0 * EPS * absf( e );
```
```suggestion
			tol = 1.0 * EPS * absf( e );
```

gunjjoshi replied:
Tested locally, we don't need this higher tolerance.

Neerajpathak07 replied:
on it.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.native.js:
```diff
@@ -0,0 +1,82 @@
+			t.equal( y, e, 'x: '+x[ i ]+'. E: '+e );
+		} else {
+			delta = absf( y - e );
+			tol = 130.0 * EPS * absf( e );
```
```suggestion
			tol = 1.0 * EPS * absf( e );
```

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

Neerajpathak07 committed:
chore: stuff from code review

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.js:
```diff
@@ -0,0 +1,79 @@
+		e = float64ToFloat32( expected[ i ] );
+		if ( y === e ) {
+			t.equal( y, e, 'x: '+x[ i ]+'. E: '+e );
+		} else {
+			delta = absf( y - e );
+			tol = 1.0 * EPS * absf( e );
+			t.ok( delta <= tol, 'within tolerance. x: '+x[ i ]+'. y: '+y+'. E: '+e+'. tol: '+tol+'. Δ: '+delta+'.' );
```
Went with low tolerance but still the test cases are failing is this something to do with the implementation.

gunjjoshi replied:
You're right. Let's keep it back to `130.0 * EPS` only.

Neerajpathak07 committed:
chore: rearrange modules

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.js:
```diff
@@ -0,0 +1,79 @@
+			t.equal( y, e, 'x: '+x[ i ]+'. E: '+e );
+		} else {
+			delta = absf( y - e );
+			tol = 1.0 * EPS * absf( e );
```
```suggestion
			tol = 130.0 * EPS * absf( e );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/test/test.native.js:
```diff
@@ -0,0 +1,88 @@
+			t.equal( y, e, 'x: '+x[ i ]+'. E: '+e );
+		} else {
+			delta = absf( y - e );
+			tol = 1.0 * EPS * absf( e );
```
```suggestion
			tol = 130.0 * EPS * absf( e );
```

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

Neerajpathak07 commented:
@gunjjoshi does this look good now??

Neerajpathak07 committed:
chore: update benchmark

Neerajpathak07 commented:
@gunjjoshi lmk if there are still some changes to be made.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/kernel-log1pf/lib/main.js:
```diff
@@ -0,0 +1,157 @@
+	t1 = float64ToFloat32( w * polyvalP( w ) );
+	t2 = float64ToFloat32( z * polyvalQ( w ) );
+	R = float64ToFloat32( t2 + t1 );
+	hfsq = float64ToFloat32( 0.5 * f * f );
```
```suggestion
	hfsq = float64ToFloat32( 0.5 * float64ToFloat32( f * f ) );
```

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/kernel-log1pf/lib/main.js

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

Neerajpathak07 commented:
@gunjjoshi Thank you for the reviews!!

Neerajpathak07 commented:
@kgryte Could we land this in?

Neerajpathak07 commented:
@kgryte could you pls take a look at this will be needing this to implement `log2f` and `log10f`

Neerajpathak07 commented:
I don't think this package had been worked upon before went through the issues and PR's to find if this had been worked upon before.
@gunjjoshi what do you think

Neerajpathak07 commented:
@kgryte Could we land this in will be needing it to implement a few other packages?

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-kernel-log1pf

anandkaranubc committed:
refactor: fix JS implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: fix C implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: update examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: update TS declarations and repl documentation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
bench: update benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
docs: update function description for clarity

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: update copyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM
